apitools/org.eclipse.pde.api.tools/src/org/eclipse/pde/api/tools/internal/model/ArchiveApiTypeContainer.java

```diff
@@ -149,31 +122,54 @@ public String toString() {
 	String fLocation;
 
 	/**
-	 * Cache of package names to class file paths in that package, or
-	 * <code>null</code> if not yet initialized.
+	 * Cache of package names to a map of class names to class files paths in that
+	 * package, or <code>null</code> if not yet initialized.
 	 */
-	private Map<String, Set<String>> fPackages;
+	private Map<String, Map<String, String>> fPackages;
 
 	/**
 	 * Cache of package names in this archive.
 	 */
 	private String[] fPackageNames;
 
 	/**
-	 * Constructs an {@link IApiTypeContainer} container for the given jar or
-	 * zip file at the specified location.
+	 * Constructs an {@link IApiTypeContainer} container for the given jar or zip
+	 * file at the specified location.
 	 *
 	 * @param parent the parent {@link IApiElement} or <code>null</code> if none
-	 * @param path location of the file in the local file system
+	 * @param path   location of the file in the local file system
 	 */
 	public ArchiveApiTypeContainer(IApiElement parent, String path) {
 		super(parent, IApiElement.API_TYPE_CONTAINER, path);
 		this.fLocation = path;
-		if (path.endsWith("jrt-fs.jar")) { //$NON-NLS-1$
-			IPath newPath = new Path(path);
-			newPath = newPath.removeLastSegments(2).addTrailingSeparator();
-			newPath = newPath.append("jmods").append("java.base.jmod"); //$NON-NLS-1$ //$NON-NLS-2$
-			this.fLocation = newPath.toOSString();
+	}
+
+	/**
+	 * Converts the location to a path in the applicable file system.
+	 *
+	 * @return the path corresponding to the location.
+	 * @throws IOException
+	 */
+	@SuppressWarnings("nls")
+	private Path getLocation() throws IOException {
+		if (fLocation.endsWith("jrt-fs.jar")) {
+			AtomicReference<IOException> exception = new AtomicReference<>();
+			FileSystem jrtFileSystem = JRTS.computeIfAbsent(fLocation, it -> {
```

Comment:
I wonder if not the String `fLocation` should be used as key here but the canonicalized/real path instead.
This should cover (the probably more edgy case) that the same target is referenced with different paths, e.g. a mixture of forward/backward slashes on Linux or Windows or the use of `.` or `..`.

If I traced the callers correctly the `ApiFileGenerationApplication` has user-defined arguments that end up here.

Reply:
Good idea.  Better to be conservative.